### PR TITLE
23 manipulate val

### PR DESCRIPTION
**Pull request by lannelin**
perform same manipulation on validation set as on training.

validation set is now passed directly to `CIFAR10DMSubset` alongside training data rather than generated again.

some bits I don't love here like the patching in conftest and the repeated `.setup()` in `test_data.py` - happy to take suggestions for changes!

### CODE REVIEW

File: tests/test_mmd.py [hunk 1]
```diff
@@ -14,13 +14,32 @@
 
 def test_cifar_mmd_same():
     dmpair = DMPair(metric_config=METRIC_CONFIG)
+    dmpair.A.setup()
+    dmpair.B.setup()
```
Comment (philswatton) [inline tests/test_mmd.py:18]: We could possibly move both `.setup()` calls inside of `dmpair.compute_similarity` (maybe with an appropriate warning/info) to avoid constantly writing it out (possibly with an argument defaulting to `True`?). This should be fine given that we'll be running two different pipelines for the model + attack training and the metrics

**Review by philswatton — CHANGES_REQUESTED** [review]
Mostly looks good to me! The main issue is that there will be an error if val_split is 0 atm because `split_indices` can't handle an empty list as input. I've included a suggestion below for handling this. I've also made some other more optional suggestions

File: src/modsim2/data/loader.py [hunk 2]
```diff
@@ -97,12 +100,7 @@ def setup(self, stage: Optional[str] = None) -> None:
             )
 
             self.dataset_train.dataset.transform = train_transforms
-            dataset_val = self.dataset_cls(
-                self.data_dir, train=True, transform=val_transforms, **self.EXTRA_ARGS
-            )
-
-            # Split
-            self.dataset_val = self._split_dataset(dataset_val, train=False)
+            self.dataset_val.dataset.transform = val_transforms
```
[the remark this answers is not in the document]
Comment (lannelin) [reply src/modsim2/data/loader.py:103]: I think this is handled in `if stage == "test"` at the moment. on l105. We might want to revisit at a later point too

File: tests/test_mmd.py (hunk 1 above)
Comment (lannelin) [reply tests/test_mmd.py:18]: yes, good idea. I'll put it in to check if already run and do this if not

File: src/modsim2/data/loader.py [hunk 3]
```diff
@@ -292,12 +294,17 @@ def __init__(
...
+            drop_percent_B=self.drop_percent_B,
+            seed=self.seed,
+            cifar=cifar,
+        )
+        val_indices_A, val_indices_B = split_indices(
+            indices=cifar.dataset_val.indices,
+            labels=[image[1] for image in cifar.dataset_val],
             drop_percent_A=self.drop_percent_A,
             drop_percent_B=self.drop_percent_B,
             seed=self.seed,
```
[the remark this answers is not in the document]
Comment (lannelin) [reply src/modsim2/data/loader.py:310]: good idea!

File: tests/test_mmd.py (hunk 1 above)
Comment (lannelin) [reply tests/test_mmd.py:18]: in fact, we don't actually need to call `setup()` at this point as we haven't implemented any transforms... I've removed this and linked to issue-15 in the code.

**Review by philswatton — APPROVED** [review]
LGTM!
